README.md
Update [2020-03-06] -- **HTTP Error 403** is sometimes responded during login process. Refering to [relevant issue on youtube-dl](https://github.com/ytdl-org/youtube-dl/issues/24008) and the corresponing [issue on this git](https://github.com/rojter-tech/pluradl.py/issues/12).
---
# Automated download from Pluralsight with `pluradl.py`
This project, `pluradl.py`, is aiming for automating the process of downloading **a whole sequence of Pluralsight courses** at once with safe parameters. Below I give an example of how to use `pluradl.py` with a Pluralsight account to get videos from an arbitrary large list of courses at their site.

**You can get a free one month trial activation code for Pluralsight by first register a [Visual Studio Dev Essentials](https://www.visualstudio.com/dev-essentials/) account for free.**

### Requirements
* [Python 3](https://www.python.org/downloads/release/python-374/)

## Install and execute
```bash
git clone https://github.com/rojter-tech/pluradl.py
cd pluradl.py
python pluradl.py
```

Dowloaded exercise files will be placed in [exercise_files](https://github.com/rojter-tech/pluradl.py/tree/master/exercise_files) directory.

## Usage

### Download from **Pluralsight** with `pluradl.py`
Make sure that [`courselist.txt`](https://github.com/rojter-tech/pluradl.py/blob/master/courselist.txt) is in the same directory as [`pluradl.py`](https://github.com/rojter-tech/pluradl.py/blob/master/pluradl.py) with the course ID's of your choice **listed row by row**. Example files and scripts is provided in [scripts](https://github.com/rojter-tech/pluradl.py/tree/master/scripts) and [scrapeutils](https://github.com/rojter-tech/pluradl.py/tree/master/scrapeutils) together with some [filtered results](https://github.com/rojter-tech/pluradl.py/tree/master/scrapeutils/filtered_results) with course IDs. The course ID can be found via the course URL from the Pluralsight website, e.g [https://app.pluralsight.com/library/courses/c-sharp-fundamentals-with-visual-studio-2015/table-of-contents](https://app.pluralsight.com/library/courses/c-sharp-fundamentals-with-visual-studio-2015/table-of-contents) where the ID is "c-sharp-fundamentals-with-visual-studio-2015".

All downloaded courses will be placed in the [courses](https://github.com/rojter-tech/pluradl.py/tree/master/courses) directory.

Run `pluradl.py` **in a terminal** to download all the videos from all the courses in [`courselist.txt`](https://github.com/rojter-tech/pluradl.py/blob/master/courselist.txt). The videos will be automatically placed in course specific directories and named by playlist order number. Substitute the example credentials with your own and supply courselist.txt with your desired courses ...

```bash
$ python pluradl.py
Enter you Pluralsight credentials
Enter username: example@somemail.com
Enter password (will not be displayed)
: mypassword
```

... with `courselist.txt` available at the same path

`courselist.txt`
```notepad
c-sharp-fundamentals-with-visual-studio-2015
csharp-nulls-working
csharp-best-practices-collections-generics
object-oriented-programming-fundamentals-csharp
using-csharp-interfaces
linq-fundamentals-csharp-6
.
.
```

#### Argument scripting
For even more automation, the script can be executed directly by passing Pluralsight username and password

```bash
python pluradl.py "example@somemail.com" "mypassword"
```
with exactly two arguments, the **first must be the username** and the **second the password**.

Or using flags

```bash
python pluradl.py --pass "mypassword" --user "example@somemail.com"
```
where the order doesn't matter.

#### Turn off subtitle
To turn off recording of subtitles use the **-s**, **--sub** or **--subtitle** flag
```bash
$ python pluradl.py --subtitle
^C
$ python pluradl.py "example@somemail.com" "mypassword" --subtitle
^C
$ python pluradl.py --user "example@somemail.com" --pass "mypassword" --subtitle
^C
```

#### Set video interval
If you only want a specific interval/range of videoclips from a specific course you can specify {the end number} or {the start and the end number} on the same row as the course id (they comes in the same order as they do on the website playlist with start number of 1)

eg. in `courselist.txt`    
```notepad
c-sharp-fundamentals-with-visual-studio-2015 25
csharp-best-practices-collections-generics 11 56
.
.
```

wich would give you the **25 first videoclips** in the course c-sharp-fundamentals-with-visual-studio-2015 and the clips **numbered 11 up to 56** in csharp-best-practices-collections-generics.

If you specify **three of more numbers** it will download those specific video indicies

`courselist.txt`    
```notepad
c-sharp-fundamentals-with-visual-studio-2015 2 10 15
csharp-best-practices-collections-generics 5 3 10 11
.
.
```

wich would give you **three videos**, those that are numbered **2**, **10** and **15**, in the course c-sharp-fundamentals-with-visual-studio-2015 and the **four specific videos** with indicies **5**, **3**, **10** and **11** in csharp-best-practices-collections-generics.

### Download exercise files (optional)
To download exercise files you need the selenium external library. Then you will need an [appropriate Chrome driver](https://sites.google.com/a/chromium.org/chromedriver/downloads) and make it availaible in system/user PATH. Selenium documentation can be [found here](https://selenium-python.readthedocs.io/).
```bash
pip install selenium
python pluraexercise.py
```

### Example output

![Directory tree of pluradl.py root](https://raw.githubusercontent.com/rojter-tech/pluradl.py/master/image/example_output_tree.png)

# IMPORTANT
The parameters `SLEEP_INTERVAL = 40` and `SLEEP_OFFSET = 120` used in the [`pluradl.py`](https://github.com/rojter-tech/pluradl.py/blob/master/pluradl.py) script is important. It means that the program will wait in average 100s (~1.7 minutes) before it downloads the next video. If you don't use these settings _Pluralsight_ will ban you because you are doing too many requests under a short period of time.

>We have blocked your account because our security systems have flagged your Pluralsight account for an unusual amount activity. This does mean a high volume of requests that are in the realm of a request every 10-30 seconds for a prolonged period of time. Please note that this high volume of activity is in violation of our terms of service [https://www.pluralsight.com/terms].

> 403    
Your account has been blocked due to suspicious activity.
Please contact support@pluralsight.com if you believe this was in error.


changelog
What's new in v1.0.0-beta3

- Fix flag parsing error (resolves [#11](https://github.com/rojter-tech/pluradl.py/issues/11)).
- Defines Pluralsight specific request headers.
- Keeps login session between playlist downloads.

pluradl.py
from __future__ import unicode_literals
import sys, os, shutil, re, getpass, io, certifi, plura_dl
from plura_dl import PluraDL
from plura_dl.utils import ExtractorError, DownloadError
if sys.version_info[0] <3:
    raise Exception("Must be using Python 3")

certpath = os.path.abspath(certifi.where())
os.environ["SSL_CERT_FILE"] = certpath

# IMPORTANT SETTINGS TO PREVENT SPAM BLOCKING OF YOUR ACCOUNT/IP AT PLURALSIGHT # # # #
SLEEP_INTERVAL = 40     # minimum sleep time        #                                 #
SLEEP_OFFSET   = 120    # adding random sleep time  #  Change this at your own risk.  #
RATE_LIMIT     = 10**6  # download rate in bytes/s  #                                 #
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

# Global defaults
DLPATH, USERNAME, PASSWORD = "", "", ""
INPROGRESSPATH, FINISHPATH, FAILPATH, INTERRUPTPATH = "", "", "", ""
PDL_OPTS = {}
SUBTITLE_OFF = False
FILENAME_TEMPLATE = r"%(playlist_index)s-%(chapter_number)s-%(title)s-%(resolution)s.%(ext)s"
PLURAURL = r"https://app.pluralsight.com/library/courses/"
SCRIPTPATH = os.path.dirname(os.path.abspath(sys.argv[0]))
COOKIEPATH = os.path.join(SCRIPTPATH, 'cookies')
COOKIEFILE = os.path.join(COOKIEPATH, 'cookies.txt')
if not os.path.exists(COOKIEPATH):
    os.mkdir(COOKIEPATH)
elif os.path.exists(COOKIEFILE):
    os.remove(COOKIEFILE)


class Logger(object):
    """Handling logging mechanism of PluraDL.
    
    Arguments:
        logpath {str} -- Path to logfile
    """
    def __init__(self,logpath):
        self.logpath = logpath
        with open(self.logpath, 'wt') as f: f.close

    def debug(self, msg):
        print(msg)
        with open(self.logpath, 'at') as f: f.write(msg+'\n')

    def warning(self, msg):
        print(msg)
        with open(self.logpath, 'at') as f: f.write(msg+'\n')

    def error(self, msg):
        print(msg)
        with open(self.logpath, 'at') as f: f.write(msg+'\n')


def set_playlist_options(digits):
    """Using appended digits in courselist.txt to set playlist option.
    
    Arguments:
        digits {[int]} -- List with playlist indicies
    """
    global PDL_OPTS

    n = len(digits)
    if n == 0:
        pass
    elif n == 1:
        print("Downloading video indicies up to",digits[0],"to")
        PDL_OPTS["playlistend"]   = digits[0]
    elif n == 2:
        print("Downloading video indicies from",digits[0],"up to and including",digits[1])
        PDL_OPTS["playliststart"] = digits[0]
        PDL_OPTS["playlistend"]   = digits[1]
    else:
        print("Downloading specific video indicies", digits)
        PDL_OPTS["playlist_items"] = ','.join([str(x) for x in digits])


def move_content(pdl, course_id, coursepath, completionpath):
    """Moves course content to its completion path.
    
    Arguments:
        pdl {PluraDL} -- PluraDL object
        course_id {str} -- [description]
        coursepath {str} -- [description]
        completionpath {str} -- Path where to store content
    """
    finalpath = os.path.join(completionpath, course_id)
    pdl.to_stdout("Moving content to " + finalpath)
    set_directory(completionpath)
    try:
        if os.path.exists(finalpath):
            shutil.rmtree(finalpath)
        shutil.move(coursepath,finalpath)
        if os.path.exists(INPROGRESSPATH):
            shutil.rmtree(INPROGRESSPATH)
    except PermissionError:
        print("Directory still in use, leaving it. Will be fixed in future releases.")


def invoke_download(course_id, course_url, coursepath):
    """Using plura_dl API to invoke download requests with associated parameters.
    
    Arguments:
        course_id {str} -- Course identifier
        course_url {str} -- Playlist url
        coursepath {str} -- Local temporary course storage path
    
    Returns:
        Bool -- Validation of completion level
    """
    with PluraDL(PDL_OPTS) as pdl:
        try:
            # Invoke download
            set_directory(coursepath)
            pdl.download([course_url])

            # Moving content to _finished destination path if the download was sucessful
            pdl.to_stdout("The course '" + course_id + "' was downloaded successfully.")
            move_content(pdl, course_id, coursepath, FINISHPATH)
            return True

        except ExtractorError:
            # Handling the case of invalid download requests
            pdl.to_stdout("The course '" + course_id + "' may not be a part of your current licence.")
            pdl.to_stdout("Visit " + course_url + " for more information.\n")
            # Moving content to _failed destination 
            move_content(pdl, course_id, coursepath, FAILPATH)
            return True
        
        except DownloadError:
            # Handling the the more general case of download error
            pdl.to_stdout("Something went wrong.")
            pdl.to_stdout("The download request for '" + course_id + "' was forced to terminate.")
            pdl.to_stdout("Double check that " + course_url)
            pdl.to_stdout("exists or that your subscription is valid for accessing its content.\n")
            # Moving content to _failed destination path
            move_content(pdl, course_id, coursepath, FAILPATH)
            return True

        except KeyboardInterrupt:
            # Handling the case of user interruption
            pdl.to_stdout("\n\nThe download stream for '" + course_id + "' was canceled by user.")
            # Moving content to _canceled destination 
            move_content(pdl, course_id, coursepath, INTERRUPTPATH)
            return False


def pluradl(course):
    """Handling the video downloading requests for a single course.
    
    Arguments:
        course {(str, [])} -- Course identifier and playlist parameters
    
    Returns:
        str -- youtue-dl CLI command
    """
    global PDL_OPTS

    # Course metadata
    course_id = course[0]
    pl_digits = course[1]
    set_playlist_options(pl_digits)
    course_url = PLURAURL + course_id

    # OS parameters - Setting up paths metadata
    coursepath = os.path.join(INPROGRESSPATH,course_id)

    # Invoking download if not already finished
    if not os.path.exists(os.path.join(FINISHPATH, course_id)):
        # Setting progress structure
        if not os.path.exists(INPROGRESSPATH):
            os.mkdir(INPROGRESSPATH)
        set_directory(coursepath)
        # Setting up logging metadata
        logfile = course_id + ".log"
        logpath = os.path.join(coursepath,logfile)
        PDL_OPTS["logger"] = Logger(logpath)
        
        return invoke_download(course_id, course_url, coursepath)
    else:
        print("Course", course_id, "already downloaded")
        return True


def flag_parser():
    """Argument handling of 4 or more arguments, interpreting arguments
    as flags with associated values.
    
    Returns:
        Bool -- Validation of argument input
    """
    if len(sys.argv) < 5:
        return False
    
    global USERNAME
    global PASSWORD

    def _check_flags(key, flag_states, arg_string=' '.join(sys.argv[1:])):
        for flag in flag_states[key][1:]:
            if flag in all_flags:
                lookaroundflag = r'(?<=' + flag + ' ' +  ')'
                lookaroundflag+=r".*?[\S]+"
        return re.findall(lookaroundflag, arg_string)

    def _check_inputs(key, user_inputs):
        for user_input in user_inputs:
            user_input = user_input.strip()
            if user_input not in (all_flags):
                flag_inputs[key] = user_input
                break # will take the first valid input
    
    usn_psw_flag_state = False
    flag_states = {"usn":[False],"psw":[False]}
    flag_inputs = {}
    username_flags = ("--user", "--username", "-u")
    password_flags = ("--pass", "--password", "-p")
    all_flags=(username_flags+password_flags)
    
    for arg in sys.argv[1:]:
        if arg in username_flags:
            flag_states["usn"][0] = True
            flag_states["usn"].append(arg)
        if arg in password_flags:
            flag_states["psw"][0] = True
            flag_states["psw"].append(arg)
    
    if flag_states["usn"][0] and flag_states["psw"][0]:
        usn_psw_flag_state = True
    
    for key in flag_states.keys():
        if flag_states[key][0]:
            user_inputs = _check_flags(key, flag_states)
            if user_inputs:
                _check_inputs(key, user_inputs)
    
    if (not "usn" in flag_inputs.keys()) or (not "psw" in flag_inputs.keys()):
        usn_psw_flag_state = False
    
    if usn_psw_flag_state:
        USERNAME = flag_inputs["usn"]
        PASSWORD = flag_inputs["psw"]
        return True
    else:
        return False


def arg_parser():
    """Handling of simple username and password argument input.
    
    Returns:
        Bool -- Validation of argument input
    """
    if len(sys.argv) < 3:
        return False
    global USERNAME
    global PASSWORD

    username = sys.argv[1]
    password = sys.argv[2]
    if username[0] != '-' and password[0] != '-':
        USERNAME = sys.argv[1]
        PASSWORD = sys.argv[2]
        return True
    else:
        return False


def get_usr_pw():
    """Requesting credentials from the user.
    
    Raises:
        ValueError: User enters an empty password too many times
    """
    print("Enter you Pluralsight credentials")
    for attempt in ["First","Second","Last"]:
        u0 = input("Enter username: ")
        if u0 == "":
            print("Username cannot be empty, enter username again")
            print(attempt, "attempt failed")
            continue
        else:
            USERNAME = u0
        
        print("Enter password (will not be displayed)")
        p0 = getpass.getpass(': ')
        if p0 != "":
            PASSWORD = p0
            return USERNAME, PASSWORD
        else:
            print('Password cannot be empty, enter password again')
    else:
        raise ValueError('Username or password was not given.')


def set_subtitle():
    """Determines whether subtitle parameters should be turned on or not.
    """
    global SUBTITLE_OFF
    subtitle_flags = ("--sub", "--subtitle", "-s",
                      "--SUB", "--SUBTITLE", "-S")
    if len(sys.argv) > 1:
        for arg in sys.argv[1:]:
            if arg in subtitle_flags:
                SUBTITLE_OFF = True
                print("Subtitles will not be appended to videoclips")


def set_directory(path):
    """Setting up directory state for os related tasks.
    
    Arguments:
        path {str} -- Full path to directory
    """
    if not os.path.exists(path):
        os.mkdir(path)
    os.chdir(path)


def get_courses(scriptpath):
    """Parsing courselist.txt separating course data.
    
    Arguments:
        scriptpath {str} -- Absolute path to script directory
    
    Returns:
        [(str, [int])] -- List of course identifiers exposed by courselist.txt
    """
    def _parse_line(line):
        course_id = ""
        digits = []

        input_chunks = re.findall(r'[\S]{1,}', line)
        for chunk in input_chunks:
            if re.search(r'[\D]{1,}', chunk):
                course_id = chunk
            else:
                digits.append(int(chunk))
        digits.sort()

        return course_id, digits
    # courses textfile prelocated inside script directory
    filelist = "courselist.txt"
    
    # Loops the list's lines and stores it as a python list
    filepath = os.path.join(scriptpath,filelist)
    courses = []
    try:
        with open(filepath, 'r+') as file:
            for line in file.readlines():
                if re.search(r'\S', line):
                    course_id, digits = _parse_line(line)
                    courses.append((course_id, digits))
        return courses
    except FileNotFoundError:
        print("There is no courselist.txt in script path. Terminating script ...")


def download_courses(courses):
    """Dowloading all courses listed in courselist.txt.
    
    Arguments:
        courses {[(str,[])]} -- List of tuples with course ID and playlist parameters.
    
    """
    # General PluraDL settings
    global PDL_OPTS
    # Globals
    PDL_OPTS["username"] = USERNAME
    PDL_OPTS["password"] = PASSWORD
    PDL_OPTS["sleep_interval"] = SLEEP_INTERVAL
    PDL_OPTS["max_sleep_interval"] = SLEEP_INTERVAL + SLEEP_OFFSET
    PDL_OPTS["ratelimit"] = RATE_LIMIT
    PDL_OPTS["outtmpl"] = FILENAME_TEMPLATE
    PDL_OPTS["cookiefile"] = COOKIEFILE
    # Tweaks
    PDL_OPTS["verbose"] = True
    PDL_OPTS["restrictfilenames"] = True
    PDL_OPTS["format"] = "bestaudio/best"
    PDL_OPTS["writesubtitles"] = True
    PDL_OPTS["allsubtitles"] = True
    PDL_OPTS["subtitlesformat"] = r'srt'
    PDL_OPTS["verbose"] = True
    if SUBTITLE_OFF:
        PDL_OPTS["writesubtitles"] = False
        PDL_OPTS["allsubtitles"] = False

    for course in courses:
        if pluradl(course):
            print("Moving to next course playlist\n")
        else:
            print("\nTerminating requests.\n")
            break


def main():
    """Main execution
    Using command line to store username and password, loops
    through the course IDs and invoking download requests.
    """
    global DLPATH, USERNAME, PASSWORD
    global INPROGRESSPATH, FINISHPATH, FAILPATH, INTERRUPTPATH

    if flag_parser():
        print("Executing by flag input ..")
    elif arg_parser():
        print("Executing by user input ..")
    else:
        USERNAME, PASSWORD = get_usr_pw()
    print("Setting username to:", USERNAME)

    set_subtitle()
    if SUBTITLE_OFF:
        subtite_state = "off"
    else:
        subtite_state = "on"
    print("Downloading subtitles is set to:", subtite_state, '\n')
    
    # Script's absolute directory path
    scriptpath = os.path.dirname(os.path.abspath(sys.argv[0]))
    
    # Download directory paths
    DLPATH = os.path.join(scriptpath,"courses")
    INPROGRESSPATH = os.path.join(DLPATH,"_inprogress")
    FINISHPATH = os.path.join(DLPATH,"_finished")
    INTERRUPTPATH = os.path.join(DLPATH,"_canceled")
    FAILPATH = os.path.join(DLPATH,"_failed")
    set_directory(DLPATH)

    # Looping through the courses determined by get_courses() invoking download requests
    courses = get_courses(scriptpath)
    if courses:
        download_courses(courses)


if __name__ == "__main__":
    main()
